File: 1016-subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer , Integer > map = new HashMap<>();
        map.put(0,1);
        int sum = 0 ; 
        int rescnt = 0 ; 
        for(int i : nums){
            sum+=i;
            int mod = sum%k; 
            if(mod<0) mod+=k ; 
            rescnt += map.getOrDefault(mod,0);
            map.merge(mod, 1  , Integer::sum);
            
        }
        return rescnt ; 
    }
}

File: 1016-subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int presum = 0 , ans = 0 ; 
        unordered_map<int , int > cntoc ; 
        cntoc[0] = 1 ; 
        for(auto it: nums){
            presum = (presum+it%k +k)%k ;
            ans+=cntoc[presum];
            cntoc[presum]++;
        }  
        return ans ;
    }
};